### PR TITLE
Improve Json perf by using JObject.Load.

### LINKED ISSUE #2783: [Perf] Use JObject.Load instead of JObject.Parse
https://github.com/dotnet/cli/blob/91fb67835d11f60bbe4763f0de6d3859236cb07a/src/dotnet/commands/dotnet-build/IncrementalCache.cs#L61

This is unnecessarily loading the string into memory. Use JObject.Load with the StreamReader.

Edit: Also here - https://github.com/dotnet/cli/blob/ac4c6702d5d345dc9c397f80290f105fedebe98b/src/Microsoft.DotNet.ProjectModel/RuntimeConfig/RuntimeConfig.cs#L40

And here - https://github.com/dotnet/cli/blob/ed05e164fbe0f046de53d7ccab343ec562bb107f/src/Microsoft.DotNet.ProjectModel/GlobalSettings.cs#L70

### PR DESCRIPTION
**Pull request by livarcocc**
Fixes #2783 and #2785 

@dotnet/dotnet-cli

### CODE REVIEW

**Comment by livarcocc** [comment]
@dotnet-bot Test Windows_NT x86 Debug Build

File: src/Microsoft.DotNet.Cli.Utils/RuntimeConfig.cs [hunk 1]
```diff
@@ -32,8 +33,10 @@ public static bool IsApplicationPortable(string entryAssemblyPath)
         } 
  
         private JObject OpenRuntimeConfig(string runtimeConfigPath) 
-        { 
-            return JObject.Parse(File.ReadAllText(runtimeConfigPath)); 
+        {
+            var reader = new JsonTextReader(new StringReader(File.ReadAllText(runtimeConfigPath)));
+
+            return JObject.Load(reader);
```
Comment (krwq) [inline src/Microsoft.DotNet.Cli.Utils/RuntimeConfig.cs]: nit: `return JObject.Parse(File.ReadAllText(runtimeConfigPath));`

File: src/Microsoft.DotNet.Cli.Utils/RuntimeConfig.cs [hunk 2]
```diff
@@ -32,8 +33,10 @@ public static bool IsApplicationPortable(string entryAssemblyPath)
         } 
  
         private JObject OpenRuntimeConfig(string runtimeConfigPath) 
-        { 
-            return JObject.Parse(File.ReadAllText(runtimeConfigPath)); 
+        {
+            var reader = new JsonTextReader(new StringReader(File.ReadAllText(runtimeConfigPath)));
```
Comment (krwq) [inline src/Microsoft.DotNet.Cli.Utils/RuntimeConfig.cs]: I'm surprised this JObject.Parse does not implement it this way if this is faster

Comment (nguerrera) [inline src/Microsoft.DotNet.Cli.Utils/RuntimeConfig.cs]: This still reads the whole string in to memory. Linked bug is suggesting reading it via StreamReader.

Comment (WildCard65) [inline src/Microsoft.DotNet.Cli.Utils/RuntimeConfig.cs]: Expanding on what @nguerrera JsonTextReader accepts any instance of a type that implements TextReader, @JamesNK (developer of Json.NET) says to use StreamReader (a class implementing TextReader) coupled to a FileStream (either opened explicitly or though StreamReader's constructor).

Comment (livarcocc) [inline src/Microsoft.DotNet.Cli.Utils/RuntimeConfig.cs]: Done.

**Comment by livarcocc** [comment]
@MattGertz for approval.

**Comment by JamesNK** [comment]
Move code that uses the JObject outside of the using statements. The yield logic especially.

Once JObject.Read is finished then you don't need to keep the file open.

**Comment by MattGertz** [comment]
@livarcocc Template?

**Comment by livarcocc** [comment]
**Customer scenario**

This is just changing the Json.NET APIs that we use to some that @JamesNK recommends as being faster.

**Bugs this fixes**

[CLI #2783](https://github.com/dotnet/cli/issues/2783)
[CLI #2785](https://github.com/dotnet/cli/issues/2785)

**Workarounds, if any**

N/A

**Risk**

Little. Just switching to a different API to load json.

**Performance impact**

Yes.

**Root cause analysis**

N/A

**How was the bug found?**

Review of the code

**Comment by livarcocc** [comment]
@MattGertz template added.

**Comment by livarcocc** [comment]
@dotnet-bot Test Ubuntu16.04 x64 Debug Build

File: src/Microsoft.DotNet.Cli.Utils/RuntimeConfig.cs [hunk 3]
```diff
@@ -31,9 +36,11 @@ public static bool IsApplicationPortable(string entryAssemblyPath)
             return false; 
         } 
  
-        private JObject OpenRuntimeConfig(string runtimeConfigPath) 
-        { 
-            return JObject.Parse(File.ReadAllText(runtimeConfigPath)); 
+        private JObject OpenRuntimeConfig(StreamReader streamReader) 
+        {
+            var reader = new JsonTextReader(streamReader);
```
Comment (wli3) [inline src/Microsoft.DotNet.Cli.Utils/RuntimeConfig.cs]: nit: why not inline, one less temp variable that could introduce bug

**Pull request merged**